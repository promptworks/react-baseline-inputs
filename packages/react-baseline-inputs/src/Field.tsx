import * as React from "react";
import { FieldProps } from "./types";
import { useTheme } from "./theme";
import { generateUniqueId, join } from "./utils";

export function Field<T extends FieldProps>({
  label,
  labelPosition = "before",
  render,
  id = React.useMemo(generateUniqueId, []),
  help,
  inline,
  touched,
  error,
  success = touched && !error,
  large,
  small,
  wrapper = true,
  theme = useTheme("field"),
  disabled,
  className,
  required,
  requiredIndicator = "*",
  ...props
}: T) {
  const fieldClassNames = [theme.field];
  const inputClassNames = [theme.input];
  const labelClassNames = [theme.label];
  const errorClassNames = [theme.error];
  const helpClassNames = [theme.help];
  const requiredIndicatorClassNames = [theme.requiredIndicator];

  if (className) {
    fieldClassNames.push(className);
  }

  if (inline) {
    fieldClassNames.push(theme.fieldInline);
    inputClassNames.push(theme.inputInline);
    labelClassNames.push(theme.labelInline);
    errorClassNames.push(theme.errorInline);
    helpClassNames.push(theme.helpInline);
  }

  if (touched) {
    fieldClassNames.push(theme.fieldTouched);
    inputClassNames.push(theme.inputTouched);
    labelClassNames.push(theme.labelTouched);
  }

  if (error) {
    fieldClassNames.push(theme.fieldError);
    inputClassNames.push(theme.inputError);
    labelClassNames.push(theme.labelError);
  }

  if (success) {
    fieldClassNames.push(theme.fieldSuccess);
    inputClassNames.push(theme.inputSuccess);
    labelClassNames.push(theme.labelSuccess);
  }

  if (large) {
    fieldClassNames.push(theme.fieldLarge);
    inputClassNames.push(theme.inputLarge);
    labelClassNames.push(theme.labelLarge);
    errorClassNames.push(theme.errorLarge);
    helpClassNames.push(theme.helpLarge);
  }

  if (small) {
    fieldClassNames.push(theme.fieldSmall);
    inputClassNames.push(theme.inputSmall);
    labelClassNames.push(theme.labelSmall);
    errorClassNames.push(theme.errorSmall);
    helpClassNames.push(theme.helpSmall);
  }

  if (disabled) {
    fieldClassNames.push(theme.fieldDisabled);
    inputClassNames.push(theme.inputDisabled);
    labelClassNames.push(theme.labelDisabled);
  }

  const labelId = `${id}_label`;
  const errorLabelId = `${id}_error`;

  const [Wrapper, wrapperProps] = wrapper
    ? ["div", { className: join(fieldClassNames) }]
    : [React.Fragment, {}];

  const inputProps = {
    id,
    disabled,
    className: join(inputClassNames),
    "aria-labelledby": error ? `${labelId} ${errorLabelId}` : undefined,
    // Adding both "aria-required" and "required" based on docs here:
    // https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required#Usability
    "aria-required": required ? "true" : undefined,
    required,
    ...props
  };

  return (
    <Wrapper {...wrapperProps}>
      {labelPosition === "after" && render(inputProps)}

      {label && (
        <label
          className={join(labelClassNames)}
          htmlFor={id}
          id={error ? labelId : undefined}
        >
          {label}
          {required && (
            <span className={join(requiredIndicatorClassNames)}>
              {requiredIndicator}
            </span>
          )}
          {help && <span className={join(helpClassNames)}>{help}</span>}
        </label>
      )}

      {labelPosition === "before" && render(inputProps)}

      {error && (
        <span role="alert" className={join(errorClassNames)} id={errorLabelId}>
          {error}
        </span>
      )}
    </Wrapper>
  );
}
