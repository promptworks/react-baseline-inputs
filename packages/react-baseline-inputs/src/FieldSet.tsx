import * as React from "react";
import { FieldSetProps, FieldSetTheme } from "./types";
import { useTheme } from "./theme";
import { generateUniqueId, join } from "./utils";

export const FieldSet: React.FC<FieldSetProps> = ({
  legend,
  id = React.useMemo(generateUniqueId, []),
  help,
  children,
  className,
  theme = useTheme("fieldSet"),
  touched,
  error,
  success = touched && !error,
  large,
  small
}) => {
  const fieldSetTheme = theme as FieldSetTheme;
  const fieldSetClassNames = [fieldSetTheme.fieldSet];
  const legendClassNames = [fieldSetTheme.legend];
  const helpClassNames = [fieldSetTheme.help];
  const errorClassNames = [fieldSetTheme.error];

  if (className) {
    fieldSetClassNames.push(className);
  }

  if (error) {
    fieldSetClassNames.push(fieldSetTheme.fieldSetError);
    legendClassNames.push(fieldSetTheme.legendError);
  }

  if (success) {
    fieldSetClassNames.push(fieldSetTheme.fieldSetSuccess);
    legendClassNames.push(fieldSetTheme.legendSuccess);
  }

  if (touched) {
    fieldSetClassNames.push(fieldSetTheme.fieldSetTouched);
    legendClassNames.push(fieldSetTheme.legendTouched);
  }

  if (large) {
    fieldSetClassNames.push(fieldSetTheme.fieldSetLarge);
    legendClassNames.push(fieldSetTheme.legendLarge);
    errorClassNames.push(fieldSetTheme.errorLarge);
    helpClassNames.push(fieldSetTheme.helpLarge);
  }

  if (small) {
    fieldSetClassNames.push(fieldSetTheme.fieldSetSmall);
    legendClassNames.push(fieldSetTheme.legendSmall);
    errorClassNames.push(fieldSetTheme.errorSmall);
    helpClassNames.push(fieldSetTheme.helpSmall);
  }

  const errorLabelId = `${id}_error`;

  return (
    <fieldset
      id={id}
      aria-describedby={error ? errorLabelId : undefined}
      className={join(fieldSetClassNames)}
    >
      <legend className={join(legendClassNames)}>{legend}</legend>

      {help && <p className={join(helpClassNames)}>{help}</p>}

      {children}

      {error && (
        <span role="alert" className={join(errorClassNames)} id={errorLabelId}>
          {error}
        </span>
      )}
    </fieldset>
  );
};
