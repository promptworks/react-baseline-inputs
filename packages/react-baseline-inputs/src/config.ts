import { Config } from "./types";
import { useContext, createContext } from "react";

export const defaultTheme = {
  field: "field",
  fieldInline: "field--inline",
  fieldSuccess: "field--success",
  fieldError: "field--erroneous",
  fieldTouched: "field--touched",
  fieldLarge: "field--large",
  fieldSmall: "field--small",
  fieldDisabled: "field--disabled",

  input: "field__input",
  inputInline: "field__input--inline",
  inputSuccess: "field__input--success",
  inputError: "field__input--erroneous",
  inputTouched: "field__input--touched",
  inputLarge: "field__input--large",
  inputSmall: "field__input--small",
  inputDisabled: "field__input--disabled",

  label: "field__label",
  labelInline: "field__label--inline",
  labelSuccess: "field__label--success",
  labelError: "field__label--erroneous",
  labelTouched: "field__label--touched",
  labelLarge: "field__label--large",
  labelSmall: "field__label--small",
  labelDisabled: "field__label--disabled",

  help: "field__help",
  helpInline: "field__help--inline",
  helpSmall: "field__help--small",
  helpLarge: "field__help--large",

  error: "field__error",
  errorInline: "field__error--inline",
  errorSmall: "field__error--small",
  errorLarge: "field__error--large"
};

export const defaultConfig: Config = {
  input: defaultTheme,
  checkbox: defaultTheme,
  dateInput: defaultTheme,
  dateTimeInput: defaultTheme,
  field: defaultTheme,
  fileInput: defaultTheme,
  floatInput: defaultTheme,
  integerInput: defaultTheme,
  radioGroup: defaultTheme,
  select: defaultTheme,
  textarea: defaultTheme,
  toggleButton: defaultTheme
};

export const ThemeContext = createContext(defaultConfig);
export const ThemeProvider = ThemeContext.Provider;

export const useTheme = (element: keyof Config) => {
  const config = useContext(ThemeContext);
  return config[element];
};
