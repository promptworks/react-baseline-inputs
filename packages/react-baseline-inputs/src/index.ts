export * from "./Checkbox";
export * from "./Field";
export * from "./DateInput";
export * from "./DateTimeInput";
export * from "./FileInput";
export * from "./FloatInput";
export * from "./IntegerInput";
export * from "./Select";
export * from "./TextArea";
export * from "./TextInput";
export * from "./types";
