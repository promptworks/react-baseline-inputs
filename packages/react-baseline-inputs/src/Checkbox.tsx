import * as React from "react";
import { CustomInputProps } from "./types";
import { Field } from "./Field";

export type CheckboxProps = CustomInputProps<"input", boolean | null>;

/**
 * An HTML `<input />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It emits a `null` value to the `onChange` handler when the input is empty.
 */
export const Checkbox: React.FC<CheckboxProps> = ({
  onChange,
  value,
  label = false,
  wrap = true,
  ...props
}) => {
  const handleChange = React.useCallback(
    event => onChange(event.target.checked),
    [onChange]
  );

  return (
    <Field
      label={label}
      render={inputProps => (
        <input
          {...inputProps}
          type="checkbox"
          checked={Boolean(value)}
          onChange={handleChange}
        />
      )}
      {...props}
    />
  );
};
