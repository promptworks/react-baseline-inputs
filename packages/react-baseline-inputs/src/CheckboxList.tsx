import * as React from "react";
import { Field } from "./Field";
import { CheckboxListProps, OptionProps } from "./types";
import { useTheme } from "./theme";

const getCheckboxProps = (option: OptionProps | string) => {
  if (typeof option === "string") {
    return { value: option, key: option, label: option };
  }

  const { value, label = value, key = value, ...props } = option;
  return { value, label, key, ...props };
};

const remove = (items: any[], item: any) => {
  const result = items.concat([]);
  result.splice(items.indexOf(item), 1);
  return result;
};

export const CheckboxList: React.FC<CheckboxListProps> = ({
  value,
  onChange,
  options = [],
  theme = useTheme("checkbox"),
  ...props
}) => {
  const values = value || [];
  const handleChange = React.useCallback(
    event => {
      const { checked, value } = event.target;

      if (checked) {
        onChange([...values, value]);
      } else {
        onChange(remove(values, value));
      }
    },
    [onChange, value]
  );

  return (
    <React.Fragment>
      {options.map(option => {
        const checkbox = getCheckboxProps(option);

        return (
          <Field
            theme={theme}
            key={checkbox.key}
            label={checkbox.label}
            labelPosition="after"
            disabled={checkbox.disabled}
            render={inputProps => (
              <input
                type="checkbox"
                checked={values.includes(checkbox.value)}
                value={checkbox.value}
                onChange={handleChange}
                {...inputProps}
              />
            )}
            {...props}
          />
        );
      })}
    </React.Fragment>
  );
};
