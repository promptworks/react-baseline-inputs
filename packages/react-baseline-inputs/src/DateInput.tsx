import * as React from "react";
import { Field } from "./Field";
import { DateInputProps } from "./types";
import { useTheme } from "./theme";

const parse = (value: string): string | null => {
  const date = new Date(value);
  const isValid = !isNaN(date.getTime());
  return isValid ? value : null;
};

/**
 * An HTML `<input type="date" />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It emits an ISO-formatted date string when changed.
 */
export const DateInput: React.FC<DateInputProps> = ({
  value,
  onChange,
  theme = useTheme("dateInput"),
  ...props
}) => {
  const handleChange = React.useCallback(
    event => onChange(parse(event.target.value)),
    [onChange]
  );

  return (
    <Field
      theme={theme}
      render={inputProps => (
        <input
          type="date"
          value={value || ""}
          onChange={handleChange}
          {...inputProps}
        />
      )}
      {...props}
    />
  );
};
