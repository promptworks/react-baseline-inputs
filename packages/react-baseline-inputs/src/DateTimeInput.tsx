import * as React from "react";
import { CustomInputProps } from "./types";
import format from "date-fns/format";
import parse from "date-fns/parse";
import isValid from "date-fns/is_valid";
import { Field } from "./Field";

export type DateTimeInputProps = CustomInputProps<"input", string | null>;

/**
 * An HTML `<input type="datetime-local" />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It emits an ISO-formatted datetime string when changed.
 */
export const DateTimeInput: React.FC<DateTimeInputProps> = ({
  value,
  onChange,
  label = false,
  wrap = true,
  ...props
}) => {
  const formattedValue = React.useMemo(
    () => (value ? format(value, "YYYY-MM-DDTHH:mm:ss") : ""),
    [value]
  );

  const handleChange = React.useCallback(
    event => {
      const value = parse(event.target.value);

      if (isValid(value)) {
        onChange(value.toISOString());
      } else {
        onChange(null);
      }
    },
    [onChange]
  );

  return (
    <Field
      label={label}
      render={inputProps => (
        <input
          {...inputProps}
          type="datetime-local"
          value={formattedValue}
          onChange={handleChange}
        />
      )}
      {...props}
    />
  );
};
