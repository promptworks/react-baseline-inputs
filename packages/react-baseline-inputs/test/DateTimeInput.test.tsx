import * as React from "react";
import { mount } from "enzyme";
import { DateTimeInput } from "../src";

describe("<DateTimeInput />", () => {
  const value = "2001-01-01T05:00:00.000Z";
  const nextLocalValue = "2018-06-13T19:00";
  const nextISOValue = "2018-06-13T23:00:00.000Z";

  it("renders with default values", () => {
    const onChange = jest.fn();
    const input = mount(<DateTimeInput value={value} onChange={onChange} />);

    expect(input).toMatchInlineSnapshot(`
      <Component
        onChange={[MockFunction]}
        value="2001-01-01T05:00:00.000Z"
      >
        <Component
          label={false}
          render={[Function]}
        >
          <div
            className="field"
          >
            <input
              className="field__input"
              id="field_1"
              onChange={[Function]}
              type="datetime-local"
              value="2001-01-01T00:00:00"
            />
          </div>
        </Component>
      </Component>
    `);
  });

  it("renders with a label", () => {
    const onChange = jest.fn();
    const input = mount(
      <DateTimeInput label="DateTime Input" value={value} onChange={onChange} />
    );

    expect(input).toMatchInlineSnapshot(`
      <Component
        label="DateTime Input"
        onChange={[MockFunction]}
        value="2001-01-01T05:00:00.000Z"
      >
        <Component
          label="DateTime Input"
          render={[Function]}
        >
          <div
            className="field"
          >
            <label
              className="field__label"
              htmlFor="field_2"
            >
              DateTime Input
            </label>
            <input
              className="field__input"
              id="field_2"
              onChange={[Function]}
              type="datetime-local"
              value="2001-01-01T00:00:00"
            />
          </div>
        </Component>
      </Component>
    `);
  });

  it("renders unwrapped input", () => {
    const onChange = jest.fn();
    const input = mount(
      <DateTimeInput wrap={false} value={value} onChange={onChange} />
    );

    expect(input).toMatchInlineSnapshot(`
      <Component
        onChange={[MockFunction]}
        value="2001-01-01T05:00:00.000Z"
        wrap={false}
      >
        <Component
          label={false}
          render={[Function]}
        >
          <div
            className="field"
          >
            <input
              className="field__input"
              id="field_3"
              onChange={[Function]}
              type="datetime-local"
              value="2001-01-01T00:00:00"
            />
          </div>
        </Component>
      </Component>
    `);
  });

  it("emits an ISO-formatted date", () => {
    const onChange = jest.fn();
    const field = mount(<DateTimeInput value={value} onChange={onChange} />);

    field.find("input").simulate("change", {
      target: { value: nextLocalValue }
    });

    expect(onChange).toHaveBeenCalledWith(nextISOValue);
  });
});
