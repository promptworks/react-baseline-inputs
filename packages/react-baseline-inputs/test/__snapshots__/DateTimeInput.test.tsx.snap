// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateTimeInput /> renders 1`] = `
<div
  class="field"
>
  <label
    class="field__label"
    for="field_1"
  >
    Jawn
  </label>
  <input
    class="field__input"
    id="field_1"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with a custom ID 1`] = `
<div
  class="field"
>
  <label
    class="field__label"
    for="foo"
  >
    Jawn
  </label>
  <input
    class="field__input"
    id="foo"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with an error 1`] = `
<div
  class="field field--erroneous"
>
  <label
    class="field__label field__label--erroneous"
    for="field_4"
    id="field_4_label"
  >
    Jawn
  </label>
  <input
    aria-labelledby="field_4_label field_4_error"
    class="field__input field__input--erroneous"
    id="field_4"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
  <span
    class="field__error"
    id="field_4_error"
    role="alert"
  >
    Oh no!
  </span>
</div>
`;

exports[`<DateTimeInput /> renders with custom requiredIndicator 1`] = `
<div
  class="field"
>
  <label
    class="field__label"
    for="field_11"
  >
    Jawn
  </label>
  <input
    class="field__input"
    id="field_11"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with help 1`] = `
<div
  class="field"
>
  <label
    class="field__label"
    for="field_5"
  >
    Jawn
    <span
      class="field__help"
    >
      Good luck
    </span>
  </label>
  <input
    class="field__input"
    id="field_5"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with inline 1`] = `
<div
  class="field field--inline"
>
  <label
    class="field__label field__label--inline"
    for="field_12"
  >
    Jawn
  </label>
  <input
    class="field__input field__input--inline"
    id="field_12"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with large 1`] = `
<div
  class="field field--large"
>
  <label
    class="field__label field__label--large"
    for="field_8"
  >
    Jawn
  </label>
  <input
    class="field__input field__input--large"
    id="field_8"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with required 1`] = `
<div
  class="field"
>
  <label
    class="field__label"
    for="field_10"
  >
    Jawn
    <span
      class="field__required-indicator"
    >
      *
    </span>
  </label>
  <input
    aria-required="true"
    class="field__input"
    id="field_10"
    required=""
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with small 1`] = `
<div
  class="field field--small"
>
  <label
    class="field__label field__label--small"
    for="field_9"
  >
    Jawn
  </label>
  <input
    class="field__input field__input--small"
    id="field_9"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with success 1`] = `
<div
  class="field field--success"
>
  <label
    class="field__label field__label--success"
    for="field_6"
  >
    Jawn
  </label>
  <input
    class="field__input field__input--success"
    id="field_6"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders with touched 1`] = `
<div
  class="field field--touched field--success"
>
  <label
    class="field__label field__label--touched field__label--success"
    for="field_7"
  >
    Jawn
  </label>
  <input
    class="field__input field__input--touched field__input--success"
    id="field_7"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders without a label 1`] = `
<div
  class="field"
>
  <input
    class="field__input"
    id="field_2"
    type="datetime-local"
    value="2001-01-01T05:00"
  />
</div>
`;

exports[`<DateTimeInput /> renders without a wrapper 1`] = `
<label
  class="field__label"
  for="field_3"
>
  Jawn
</label>
`;
